src/NHibernate/Cache/UpdateTimestampsCache.cs

```diff
@@ -72,45 +72,107 @@ public virtual void Invalidate(IReadOnlyCollection<string> spaces)
 			//TODO: to handle concurrent writes correctly, the client should pass in a Lock
 			long ts = updateTimestamps.NextTimestamp();
 			//TODO: if lock.getTimestamp().equals(ts)
-			foreach (var space in spaces)
+			if (log.IsDebugEnabled())
+				log.Debug("Invalidating spaces [{0}]", StringHelper.CollectionToString(spaces));
+			SetSpacesTimestamp(spaces, ts);
+		}
+
+		private void SetSpacesTimestamp(IReadOnlyCollection<string> spaces, long ts)
+		{
+			if (_batchUpdateTimestamps != null)
 			{
-				log.Debug("Invalidating space [{0}]", space);
-				updateTimestamps.Put(space, ts);
+				if (spaces.Count == 0)
+					return;
```

Comment:
Avoid an undue cache `PutMany` hit: NHibernate may ask to "invalidate" no spaces at all. With single operations, this is resolved to an empty loop, so no actual call to the cache is issued. With multi-operations, a multi-put call would be issued on the cache, with an empty list of keys: better avoid that call.